### PR TITLE
RSARTE-176569: Removed Poco dependency from lib-tcp-server

### PR DESCRIPTION
**Pull request by rajneshkumarmalik**
Removed POCO dependency form lib-tcp-server

### CODE REVIEW

File: LibTCPServer.emx [hunk 1]
```diff
@@ -51,7 +51,7 @@
           <element xsi:nil="true"/>
           <layoutConstraint xmi:type="notation:Bounds" xmi:id="_DUwG8-5UEeitL-zKjjNp2A" x="889" y="2159" width="5607" height="7494"/>
         </children>
-        <children xmi:type="notation:Shape" xmi:id="_kdX4MO5VEeitL-zKjjNp2A" type="Note" fontName="Segoe UI" fontHeight="8" description="&lt;p>Library for letting a Model RealTime application communicate over TCP with other applications. For example, external applications can use it to send events into the Model RealTime application. Note that this library uses the POCO library, so you must first download and build it before you can build this library.&lt;/p>" fillColor="13956346" transparency="0" lineColor="12632258" lineWidth="1">
+        <children xmi:type="notation:Shape" xmi:id="_kdX4MO5VEeitL-zKjjNp2A" type="Note" fontName="Segoe UI" fontHeight="8" description="&lt;p>Library for letting an DevOps Model RealTime application communicate over TCP with other applications. For example, external applications can use it to send events into the DevOps Model RealTime application.&lt;/p>" fillColor="13956346" transparency="0" lineColor="12632258" lineWidth="1">
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: You can omit "DevOps" to make the product name shorter. If you keep it, it should be "a" instead of "an".

File: LibTCPServer.emx [hunk 2]
```diff
@@ -341,38 +345,49 @@ srv->start();
...
 	// Everything else is handled as TCP requests
-	Parser parser;
-	Var result;
-	try {
-		result = parser.parse(str);
+	RTJsonParser parse;
+	RTJsonResult jsonResult;
+	if(!parse.parseJsonString(jsonResult, str))
+	{
+		conn->replyHTTPStatus(ITCPConnection::Status::Error, "Failed to parse JSON: " + str);
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: I think we should only reply with HTTP status if we get an HTTP request (as above). For a regular TCP request a HTTP reply will just look unreadable due to lots of extra information that we don't need.

File: LibTCPServer.emx [hunk 3]
```diff
@@ -464,22 +469,13 @@ srv->start();
...
-	    try {
-	    	Var var_portIndex = object->get("portIndex");
-	    	if (!var_portIndex.isEmpty())
-	    		portIndex = var_portIndex.convert<int>();
-	    }
-	    catch (Poco::Exception e) {
-	    	conn->replyStatus(ITCPConnection::Status::Error, "Illegal portIndex: " + e.message());
-	    	return;
-	    }
+	    int portIndex = jsonResult["portIndex"].get_int();
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Missing error handling. What if the user doesn't specify "portIndex" as an integer?

File: LibTCPServer.emx [hunk 4]
```diff
@@ -464,22 +469,13 @@ srv->start();
...
 	    
 	    if (cmd == "sendEvent") {	    
-	    	Var var_priority = object->get("priority");
+	        std::string priorityString =  jsonResult["priority"].get_string();
```
Comment (mattias-mohlin) [inline LibTCPServer.emx:497]: Missing error handling. What if the user doesn't specify "priority" as a string?

File: LibTCPServer.emx [hunk 5]
```diff
@@ -670,65 +663,83 @@ str.append("}");
...
+				<< config.remoteHost << " on port " << config.remotePort
+				<< std::endl;
+		tcp.close();
+		return;
+	}
 	writeSuccess = true;
 
+	static const RTTimespec awhile(1L, 0L);
+	while (!ioMonitor.wait(&awhile)) {
+		std::this_thread::sleep_for(std::chrono::seconds(1));
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: What do we gain by two waits? And what will happen if no response at all comes for the request? Will we get stuck in an infinite loop?

File: LibTCPServer.emx [hunk 6]
```diff
@@ -943,6 +953,13 @@ catch (Poco::Exception& exc) {
           <type xmi:type="uml:PrimitiveType" href="platform:/plugin/com.ibm.xtools.umldt.rt.cpp.core/libraries/CppPrimitiveDatatypes.emx#_f0875RKyEdyaM9S2zG4bgA?CppPrimitiveDatatypes/bool?"/>
         </defaultValue>
       </ownedAttribute>
+      <ownedAttribute xmi:id="_rXaeQPNBEe6D8ogtBjw-1g" name="maxSupportedClient" visibility="public" isStatic="true">
```
Comment (mattias-mohlin) [inline LibTCPServer.emx:993]: We must remember to mention this new configuration parameter in the README for the GitHub repo.

File: LibTCPServer.emx [hunk 7]
```diff
@@ -1050,152 +1065,146 @@ catch (Poco::Exception& exc) {
...
 		}
 		
-		try {
-			socket().sendBytes(msg.data(), (int) msg.length());
-			socket().sendBytes(NL.data(), (int) NL.length());
+		if (logToStdOut)
+			std::cout << "TCP request received: ";
+
+		// Wait for data to be available on the socket
+		static const RTTimespec awhile(5L, 0L);
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Was this 5s limit also hard-coded before? If we need it, we should at least check logToStdOut before printing status messages to stdout.

File: LibTCPServer.emx [hunk 8]
```diff
@@ -1212,35 +1221,96 @@ public:
...
+		listenSock.setMaxPendingConnections(maxSupportedClient);
+		if (!listenSock.create() || !listenSock.listen(port))
+		{
+			std::cerr << "Failed to create TCP accept socket" << std::endl;
+			listenSock.close();
+			return;
+		}
+
+		do {
+			static const RTTimespec awhile(5L /* s */, 0L/*maxWaitForReply*1000*1000*/ /* ns */);
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: I did some experiments with making this wait for maxWaitForReply, since we should respect the value given by the user for this setting. Now it seems we don't.

File: LibTCPServer.emx [hunk 9]
```diff
@@ -1212,35 +1221,96 @@ public:
...
+				}
+
+				connections.push_back(lsc);
+			}
+			else {
+				// Ongoing connections should all have been replied to by now
+				while (!connections.empty()) {
+					LocalServerConnection* lsc = connections.front();
+					if (lsc->hasReplied) {
+						std::cout << "Reply detected. Removing LocalServerConnection" << std::endl;
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: This printout was for internal use. At least we should check if user wants messages printed on stdout.

File: README.md [hunk 10]
```diff
@@ -16,25 +16,17 @@ It's possible to implement a custom handling of incoming messages, but the defau
...
 
 `
-tc.aPocoLoc = '$(RSA_RT_HOME)/tcpserver/poco/';
+tc.compileArguments = topTC.compileArguments || '$(DEBUG_TAG)';
 `
 
-List of linked libraries is defined in `tc.libraries` property of the TC `pocoLib.tcjs` based on the target configuration of the top executable.
-
-If you want to be able to build the library TC directly (i.e. as a "top TC"), you should use `tcpServerLib_custom.tcjs` and first modify the target configuration and compiler settings.
+If you want to be able to build the library TC directly (i.e. as a "top TC"), you should first modify these TC properties (essentially making sure that the value after the '||' is appropriate).
```
Comment (mattias-mohlin) [inline README.md]: Will we remove tcpServerLib_custom.tcjs now, or why did you remove the text about it that Elena wrote for last sprint? I think we need to discuss in a scrum how the user should build the library as a top TC.

**Review by mattias-mohlin — CHANGES_REQUESTED** [review]
See a few comments above.

File: LibTCPServer.emx (hunk 1 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx]: done.

File: LibTCPServer.emx (hunk 3 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx]: done.

File: LibTCPServer.emx (hunk 4 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx:497]: It is handled in below code.

File: LibTCPServer.emx (hunk 5 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx]: I used maxWaitForReply configuration and return if response is not coming within time.

File: LibTCPServer.emx (hunk 6 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx:993]: updated in readme file.

File: LibTCPServer.emx (hunk 7 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx]: Fixed by using config parameters.

File: LibTCPServer.emx (hunk 8 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx]: fixed by using config parameter maxWaitForReply.

File: LibTCPServer.emx (hunk 9 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx]: Fixed by using logToStdOut config parameter.

File: README.md (hunk 10 above)
Comment (rajneshkumarmalik) [reply README.md]: Updated readme file again and mentioned about master branch and release Tag branch.

File: LibTCPServer.emx [hunk 11]
```diff
@@ -341,38 +345,49 @@ srv->start();
...
-	Var var_cmd = object->get("command");
-	std::string cmd = var_cmd.convert<std::string>();
+	std::string cmd = jsonResult["command"].get_string();
 	if (config.logToStdOut)
 		std::cout << "JSON command received: " << cmd << std::endl;
 
 	if (cmd == "sendEvent" || cmd == "invokeEvent") {
-		Var var_port = object->get("port");
-		std::string portName = var_port.convert<std::string>();
+		std::string portName =  jsonResult["port"].get_string();
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Missing check that the "port" key is present.

File: LibTCPServer.emx [hunk 12]
```diff
@@ -397,8 +412,7 @@ srv->start();
 		const RTProtocolDescriptor* protocolDescriptorConjugate = protocolDescriptor->conjugate;
 
 		// Find a matching out event in the port protocol
-		Var var_event = object->get("event");
-		std::string eventName = var_event.convert<std::string>();
+		std::string eventName =  jsonResult["event"].get_string();
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Missing check that the "event" key is present.

File: LibTCPServer.emx [hunk 13]
```diff
@@ -465,21 +470,19 @@ srv->start();
...
-	    		portIndex = var_portIndex.convert<int>();
-	    }
-	    catch (Poco::Exception e) {
-	    	conn->replyStatus(ITCPConnection::Status::Error, "Illegal portIndex: " + e.message());
-	    	return;
+	    std::string portIndexString = jsonResult["portIndex"].get_string();
+	    if (!portIndexString.empty()) {
+            portIndex = RTMemoryUtil::atoi(portIndexString.c_str());
+	    } else {
+            conn->replyStatus(ITCPConnection::Status::Error, "Illegal portIndex: ");
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: The portIndex is missing in the error message.

File: LibTCPServer.emx [hunk 14]
```diff
@@ -670,65 +670,88 @@ str.append("}");
...
+				<< std::endl;
+		tcp.close();
+		return;
+	}
 	writeSuccess = true;
 
+	static const RTTimespec awhile(config.maxWaitForReply / 1000, 0L);
+	while (!ioMonitor.wait(&awhile)) {
+		if(config.logToStdOut) {
+			std::cout << "No out message available within 5 seconds " << std::endl;
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: 5 should be replaced with config.maxWaitForReply

File: LibTCPServer.emx [hunk 15]
```diff
@@ -670,65 +670,88 @@ str.append("}");
...
-	
+
 	if (config.logToStdOut)
 		std::cout << "Response contained JSON command: " << cmd << std::endl;
 
 	if (cmd == "reply") {
-		Var var_port = object->get("port");
-		std::string portName = var_port.convert<std::string>();
-
+		std::string portName =  jsonResult["port"].get_string();
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Missing check that the "port" key is present.

File: LibTCPServer.emx [hunk 16]
```diff
@@ -752,8 +775,7 @@ catch (Poco::Exception& exc) {
 		const RTProtocolDescriptor* protocolDescriptorConjugate = protocolDescriptor->conjugate;
 
 		// Find a matching out event in the port protocol
-		Var var_event = object->get("event");
-		std::string eventName = var_event.convert<std::string>();
+		std::string eventName =  jsonResult["event"].get_string();
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Missing check that the "event" key is present.

File: LibTCPServer.emx [hunk 17]
```diff
@@ -1050,152 +1077,152 @@ catch (Poco::Exception& exc) {
...
-			return;
+	~LocalServerConnection() {
+		clientSock.close();
+	}
+
+	bool init() {
+
+		if (!clientSock.acceptFrom(listenSock)) {
+			if(logToStdOut) {
+				std::cerr << "Connection failed to accept socket" << std::endl;
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Inconsistency. Flag checks if user wants to log to stdout, but message is written to stderr. If this is an unexpected error message, we should write to stderr without checking "logToStdOut" (as I think is already done in some other places).

File: LibTCPServer.emx [hunk 18]
```diff
@@ -1050,152 +1077,152 @@ catch (Poco::Exception& exc) {
...
-			socket().sendBytes(msg.data(), (int) msg.length());
-			socket().sendBytes(NL.data(), (int) NL.length());
+		if (logToStdOut)
+			std::cout << "TCP request received: ";
+
+		// Wait for data to be available on the socket
+		static const RTTimespec awhile(maxWaitForReply/1000, 0L);
+		while (!ioMonitorClient.wait(&awhile)) {
+			if (logToStdOut) {
+				std::cout << "No data available within 5 seconds. " << std::endl;
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Replace hardcoded 5

File: LibTCPServer.emx [hunk 19]
```diff
@@ -1212,35 +1239,100 @@ public:
...
+		enter( RTJob::running );
+
+		RTIOMonitor ioMonitor;
+		RTTcpSocket listenSock;
+		listenSock.registerWith(&ioMonitor);
+		listenSock.setMaxPendingConnections(maxSupportedClient);
+		if (!listenSock.create() || !listenSock.listen(port))
+		{
+			if(logToStdOut) {
+				std::cerr << "Failed to create TCP accept socket" << std::endl;
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Inconsistency between stdout and stderr. See above.

File: README.md [hunk 20]
```diff
@@ -1,6 +1,6 @@
 # lib-tcp-server
 A library for [DevOps Model RealTime](https://www.hcl-software.com/devops-model-realtime) which allows a realtime application to communicate over TCP with other applications.
-Note: This library requires Model RealTime 10.3 2018.48 or later, and further documentation can be found [here](https://model-realtime.hcldoc.com/help/topic/com.ibm.xtools.rsarte.webdoc/Articles/Integrations/TCPServer%20Library.html).
+Note: Master branch of this library requires Model RealTime 12.0.2 release or later while release tag V1.0 requires Model RealTime 10.3 2018.48 or later, and further documentation can be found [here](https://model-realtime.hcldoc.com/help/topic/com.ibm.xtools.rsarte.webdoc/Articles/Integrations/TCPServer%20Library.html).
```
Comment (mattias-mohlin) [inline README.md]: I suggest we rename the label to "Poco" instead, and also mentions that it requires the Poco library (in addition to Model RealTime 10.3 2018.48 or later)

File: README.md [hunk 21]
```diff
@@ -16,6 +16,22 @@ It's possible to implement a custom handling of incoming messages, but the defau
...
+
+To make use of this library add the TC `tcpServerLib.tcjs` as a prerequisite of the executable TC for your Model RealTime application.
+
+Note that the TC `tcpServerLib.tcjs` is designed to be built as a prerequisite of an executable TC and will reuse some of the TC properties from that executable (using the topTC variable). Here is an example of such a TC property:
+
+`
+tc.compileArguments = topTC.compileArguments || '$(DEBUG_TAG)';
+`
+
+If you want to be able to build the library TC directly (i.e. as a "top TC"), you should first modify these TC properties (essentially making sure that the value after the '||' is appropriate).
```
Comment (mattias-mohlin) [inline README.md]: Obsolete sentence. Now the user should instead use the other TC for building the library directly.

File: README.md [hunk 22]
```diff
@@ -16,6 +16,22 @@ It's possible to implement a custom handling of incoming messages, but the defau
 To implement your own custom handling of received messages, override the operation `TCPServer::handleReceivedMessageCustom()` and set the config property `defaultHandlingOfReceivedMessages` to false.
 
 ## Build the Library
+
+### Master branch
```
Comment (mattias-mohlin) [inline README.md]: Suggested change:
The original implementation of the library uses the POCO library and requires Model RealTime 10.3 2018.48 or later. Checkout the tag "poco" if you want to use this version.

The latest implementation, that is available on the master branch, has no dependency on POCO, but instead requires Model RealTime 12.0.2 or later.

File: README.md [hunk 23]
```diff
@@ -16,6 +16,22 @@ It's possible to implement a custom handling of incoming messages, but the defau
...
+Note that the TC `tcpServerLib.tcjs` is designed to be built as a prerequisite of an executable TC and will reuse some of the TC properties from that executable (using the topTC variable). Here is an example of such a TC property:
+
+`
+tc.compileArguments = topTC.compileArguments || '$(DEBUG_TAG)';
+`
+
+If you want to be able to build the library TC directly (i.e. as a "top TC"), you should first modify these TC properties (essentially making sure that the value after the '||' is appropriate).
+
+### Tag V1.0
+The release Tag V1.0 of this library depends on POCO library and below instruction must be follow to use this release.
```
Comment (mattias-mohlin) [inline README.md]: You don't need to document how to use the Poco version, since that's already documented in the README the user will see when he checks out that branch.

File: README.md [hunk 24]
```diff
@@ -48,7 +64,9 @@ The following configuration properties are available:
...
 If true, all incoming TCP messages are assumed to be string-encoded JSON objects according to the format described below. Set to false if you wish to implement your own custom handling of incoming messages.
 - **maxWaitForReply**:integer
-The max number of milliseconds to wait for the Model RealTime application to reply to a request for sending or invoking an event to it. If the TCPServer doesn't get a confirmation from the application within the specified time limit, it will assume that the Model RealTime application failed to receive the event and return an error message. 
+The max number of milliseconds to wait for the Model RealTime application to reply to a request for sending or invoking an event to it. If the TCPServer doesn't get a confirmation from the application within the specified time limit, it will assume that the Model RealTime application failed to receive the event and return an error message.
+- **maxSupportedClient**:integer
+The maximum number of supported client. This configuration is not available on POCO based release while available in master branch. This configuration requires Model RealTime release 12.0.2 or later.
```
Comment (mattias-mohlin) [inline README.md]: client -> clients

No need to document the POCO version since that's already done in the README the user will see when he checks out the Poco tag.

**Review by mattias-mohlin — CHANGES_REQUESTED** [review]
See comments above

File: LibTCPServer.emx [hunk 25]
```diff
@@ -51,7 +51,7 @@
           <element xsi:nil="true"/>
           <layoutConstraint xmi:type="notation:Bounds" xmi:id="_DUwG8-5UEeitL-zKjjNp2A" x="889" y="2159" width="5607" height="7494"/>
         </children>
-        <children xmi:type="notation:Shape" xmi:id="_kdX4MO5VEeitL-zKjjNp2A" type="Note" fontName="Segoe UI" fontHeight="8" description="&lt;p>Library for letting a Model RealTime application communicate over TCP with other applications. For example, external applications can use it to send events into the DevOps Model RealTime application.&lt;/p>" fillColor="13956346" transparency="0" lineColor="12632258" lineWidth="1">
+        <children xmi:type="notation:Shape" xmi:id="_kdX4MO5VEeitL-zKjjNp2A" type="Note" fontName="Segoe UI" fontHeight="8" description="&lt;p>Library for letting an DevOps Model RealTime application communicate over TCP with other applications. For example, external applications can use it to send events into the DevOps Model RealTime application.&lt;/p>" fillColor="13956346" transparency="0" lineColor="12632258" lineWidth="1">
```
Comment (mattias-mohlin) [inline LibTCPServer.emx]: Replace "an DevOps" with "a" as it was before.

**Review by mattias-mohlin — APPROVED** [review]
Ok (but see one final small comment above)

File: LibTCPServer.emx (hunk 25 above)
Comment (rajneshkumarmalik) [reply LibTCPServer.emx]: Fixed
